packages/@react-stately/table/src/utils.ts

```diff
@@ -0,0 +1,113 @@
+        'defaulting to \'1fr\'');
+    return 1;
+  }
+  return parseInt(match[0], 10);
+}
+
+export function parseStaticWidth(width: number | string, tableWidth: number): number {
+  if (typeof width === 'string') {
+    let match = width.match(/^(\d+)(?=%$)/);
+    if (!match) {
+      throw new Error('Only percentages or numbers are supported for static column widths');
+    }
+    return tableWidth * (parseInt(match[0], 10) / 100);
+  }
+  return width;
+}
+    
+    
+export function getMaxWidth(maxWidth: number | string, tableWidth: number): number {
+  return maxWidth != null
+        ? parseStaticWidth(maxWidth, tableWidth)
+        : Infinity;
+}
+
+export function getMinWidth(minWidth: number | string, tableWidth: number): number {
+  return minWidth != null
+      ? parseStaticWidth(minWidth, tableWidth)
+      : 75;
+}
+
+function mapDynamicColumns<T>(remainingColumns: GridNode<T>[], remainingSpace: number, tableWidth: number): mappedColumn<T>[] {
+  let remainingFractions = remainingColumns.reduce(
+        (sum, column) => sum + parseFractionalUnit(column.props.defaultWidth),
+        0
+      );
+    
+  let columns = [...remainingColumns].map((column, index) => {
+    const targetWidth =
+          (parseFractionalUnit(column.props.defaultWidth) * remainingSpace) / remainingFractions;
+    
+    return {
+      ...column,
+      index,
+      delta: Math.max(
+            0,
+            getMinWidth(column.props.minWidth, tableWidth) - targetWidth,
+            targetWidth - getMaxWidth(column.props.maxWidth, tableWidth)
+          )
+    };
+  });
+    
+  return columns;
+}
+
+function findDynamicColumnWidths<T>(remainingColumns: mappedColumn<T>[], remainingSpace: number, tableWidth: number): mappedColumn<T>[] {
+  let remainingFractions = remainingColumns.reduce(
+        (sum, col) => sum + parseFractionalUnit(col.props.defaultWidth),
+        0
+      );
+    
+  for (let i = 0; i < remainingColumns.length; i++) {
+    const column = remainingColumns[i];
+    
+    const targetWidth =
+          (parseFractionalUnit(column.props.defaultWidth) * remainingSpace) / remainingFractions;
+    
+    let width = Math.max(
+          getMinWidth(column.props.minWidth, tableWidth),
+          Math.min(Math.floor(targetWidth), getMaxWidth(column.props.maxWidth, tableWidth))
+        );
+    column.calculatedWidth = width;
+    remainingSpace -= width;
+    remainingFractions -= parseFractionalUnit(column.props.defaultWidth);
+  }
+    
+  return remainingColumns;
+}  
+    
+export function getDynamicColumnWidths<T>(remainingColumns: GridNode<T>[], remainingSpace: number, tableWidth: number) {
+  let columns = mapDynamicColumns(remainingColumns, remainingSpace, tableWidth);
+    
+  columns.sort((a, b) => b.delta - a.delta);
+  columns = findDynamicColumnWidths(columns, remainingSpace, tableWidth);
+  columns.sort((a, b) => a.index - b.index);
```

Comment:
I think you were worried about sort times and whatnot, so I had a thought
I believe this is (2n [reduce + map] + nlogn (sort) + 2n [reduce + map] + nlogn (sort)) => 4n + 2nlogn
but you could also keep a priority queue (heaps work well) while doing mapDynamicColumns, as you come across columns where the delta is 0, you handle them, if it's > 0 you add it to the priority queue, once you've run through the list, you begin handling the queue
handle means do what you do in findDynamicColumnWidths

I believe this would become
(n [reduce] + nlog(n) [priority queue insertion] + n [anything in priority queue]) => 2n + nlogn
so in terms of extremely large sets, pretty much equivalent, but since this isn't extremely large, this might save a bit of time

this all hinges on me understanding the above functions well enough :) and you maybe be able to simplify the priority queue further, I don't quite know how that part factors in, but if it's just a bucket where we put everything that's going to be snapped to their bounds, then maybe this could be further reduced to O(n) 3n or linear

Reply:
Hopefully this explains why we need a sort and not just a queue. The order of the delta values is important. We could potentially put all the non-zero delta values into a separate array and only sort them and then concat all the 0 delta values onto the end. This would make it so we are only sorting clamped columns which should be a very small subset. In practice though, I don't think this would actually buy us much.
Question would be is the slightly more complex code worth the performance differece?